test_util/launch.py

```diff
@@ -1,19 +1,145 @@
-"""
+"""DC/OS Launch
 
 Usage:
-dcos-launch create [--wait] [--dump-info=cluster_info.json] <config=config.yaml>
-dcos-launch wait <path=cluster_info.json>
-dcos-launch describe <path=cluster_info.json>
-dcos-launch delete <path=cluster_info.json>
-"""
+  dcos-launch create [--info-path=<path>] [--config-path=<path>]
+  dcos-launch wait [--info-path=<path>]
+  dcos-launch (describe|delete) [--info-path=<path>]
 
+Commands:
+  create    reads the file given by --config-path, creates the cluster described therein
+            and finally dumps a JSON file to the path given in --info-path which can then
+            be used with the wait, describe, and delete calls
+  wait      Block until the cluster is up and running
+  describe  Return additional information about the composition of the cluster
+  delete    Destroying the provided cluster deployment
+
+Options:
+  --config-path=<path>  Path for config to create cluster from [default: config.yaml].
+  --info-path=<path>    JSON file output by create and consumed by wait, describe,
+                        and delete [default: cluster_info.json].
+"""
+import abc
+import json
+import os
+import sys
+import yaml
 from docopt import docopt
+from pkgpanda.util import load_json, load_string, write_json
+
+from test_util.aws import BotoWrapper, DcosCfSimple
+
+
+class AbstractLauncher(metaclass=abc.ABCMeta):
+    def __init__(self):
+        raise NotImplementedError()
+
+    def create(self):
+        raise NotImplementedError()
+
+    def wait(self):
+        raise NotImplementedError()
+
+    def describe(self):
+        raise NotImplementedError()
+
+    def delete(self):
+        raise NotImplementedError()
+
+
+class AwsLauncher(AbstractLauncher):
+    def __init__(self, aws_region, aws_access_key_id, aws_secret_access_key, config):
+        self.boto_wrapper = BotoWrapper(aws_region, aws_access_key_id, aws_secret_access_key)
+        self.config = config
+
+    def create(self):
+        self.boto_wrapper.create_stack(
+            self.config['stack_name'], self.config['template_url'], self.config['parameters'])
+        return {
+            'cloudformation': {
+                'stack_name': self.config['stack_name'],
+                'region': self.config['region'],
+                'access_key_id': self.config['access_key_id'],
+                'secret_access_key': self.config['secret_access_key']}}
+
+    def wait(self):
+        # TODO: should this support the case where the cluster is being updated?
+        cf = self.get_stack()
+        status = cf.get_stack_details()['StackStatus']
+        if status == 'CREATE_IN_PROGRESS':
+            cf.wait_for_stack_creation(wait_for_poll_min=0)
+        elif status == 'CREATE_COMPLETE':
+            pass
+        else:
+            raise Exception('')
+        print('Cluster is ready!')
+
+    def describe(self):
+        cf = self.get_stack()
+        return {
+            'masters': convert_host_list(cf.get_master_ips()),
```

Comment:
How does this work if it's a blind cloudformation template? (I see this assumes AWS CF Simple, but would be better not to for the generic AWS CloudFormation Launcher).